app/src/shared/utils/withWeb3.jsx

```diff
@@ -0,0 +1,102 @@
+// @flow
+
+import React, { type ComponentType } from 'react'
+
+import getWeb3, { validateWeb3 } from '$shared/web3/web3Provider'
+import Web3ErrorDialog from '$shared/components/Web3ErrorDialog'
+import { WalletLockedError } from '$shared/errors/Web3'
+
+import Web3Poller from '$shared/web3/web3Poller'
+import type { Props as DialogProps } from '$shared/components/Dialog'
+
+type State = {
+    checkingWeb3: boolean,
+    web3Error: ?Error,
+}
+
+type OwnProps = {
+    requireWeb3: boolean,
+    onCancel?: () => void,
+}
+
+type Props = DialogProps & OwnProps
+
+export function withWeb3(WrappedComponent: ComponentType<any>) {
+    class WithWeb3 extends React.Component<Props, State> {
+        static defaultProps = {
+            requireWeb3: true,
+        }
+
+        state = {
+            checkingWeb3: false,
+            web3Error: null,
+        }
+
+        componentDidMount() {
+            if (this.props.requireWeb3) {
+                Web3Poller.subscribe(Web3Poller.events.ACCOUNT_ERROR, this.setLocked)
+                Web3Poller.subscribe(Web3Poller.events.ACCOUNT, this.validateWeb3)
+                this.validateWeb3()
+            }
+        }
+
+        componentWillUnmount() {
+            if (this.props.requireWeb3) {
+                Web3Poller.unsubscribe(Web3Poller.events.ACCOUNT_ERROR, this.setLocked)
+                Web3Poller.unsubscribe(Web3Poller.events.ACCOUNT, this.validateWeb3)
+            }
+        }
+
+        setLocked = () => {
+            this.setState({
+                checkingWeb3: false,
+                web3Error: new WalletLockedError(),
+            })
+        }
+
+        validateWeb3 = () => {
+            if (!this.props.requireWeb3 || this.state.checkingWeb3) {
+                return
+            }
+            this.setState({
+                checkingWeb3: true,
+                web3Error: null,
+            }, () => {
+                validateWeb3(getWeb3())
+                    .then(() => {
```

Comment:
when using async + setState always need to check whether component has unmounted, otherwise React prints annoying "trying to set state on unmounted component" console warning